domain/application/service/application.go

```diff
@@ -673,8 +678,27 @@ func (s *Service) MarkApplicationDead(ctx context.Context, appName string) error
 
 // SetApplicationCharm sets a new charm for the application, validating that aspects such
 // as storage are still viable with the new charm.
-func (s *Service) SetApplicationCharm(ctx context.Context, name string, params UpdateCharmParams) error {
-	//TODO(storage) - update charm and storage directive for app
+func (s *Service) SetApplicationCharm(ctx context.Context, appName string, params application.UpdateCharmParams) error {
+	appID, err := s.st.GetApplicationIDByName(ctx, appName)
+	if err != nil {
+		return errors.Errorf("getting application ID: %w", err)
+	}
+	err = s.st.SetApplicationCharm(ctx, appID, params)
+	var (
+		notImplementedRelation *internal.NotImplementedRelationError
+		quotaLimitExceed       *internal.RelationQuotaLimitExceededError
+	)
+	switch {
+	case errors.As(err, &quotaLimitExceed):
+		return errors.Errorf("new charm version imposes a maximum relation limit of %d for %s:%s which cannot be satisfied by the number of already established relations (%d)",
+			quotaLimitExceed.Relation.Limit, appName, quotaLimitExceed.Relation.Name, quotaLimitExceed.Count)
+	case errors.As(err, &notImplementedRelation):
+		return errors.Errorf("would break relation %s:%s", appName, notImplementedRelation.Relation.Name)
+	}
```

Comment:
Why do these errors have to be defined here and not in state? looks like we're mapping the original error.

Reply:
(I answer here because the call site is more interesting than the error definition). 
So, my problem with this approach is that you haven't defined any new error in the domain errors, and you return an untyped error. I'm guessing we would like to check for a "RelationQuotaLimitExceeded" error (for example) on the call sites (facade/worker). And at the same time you've defined an error type just to carry some results from state to the service layer.
Do you absolutely need the application name in the error string? Can't it be replaced by the appID? I ask because what I'd prefer here is a typed error that gets returned from the state layer, as we do with some other errors (e.g. `UnitNotAlive`).